src/main/java/com/atlassian/bitbucket/jenkins/internal/applink/oauth/serviceprovider/token/PersistentServiceProviderTokenStore.java

```diff
@@ -0,0 +1,422 @@
+            addNode(writer, VERIFIER, encrypt(token.getVerifier()));
+            addNode(writer, CALLBACK, token.getCallback());
+            addNode(writer, TOKEN_CREATION_TIME, token.getCreationTime());
+            addNode(writer, TOKEN_TIME_TO_LIVE, token.getTimeToLive());
+            addProperties(writer, token.getProperties());
+            addSession(writer, token.getSession());
+        }
+
+        @Override
+        public Object unmarshal(HierarchicalStreamReader reader, UnmarshallingContext context) {
+            Boolean accessToken = null;
+            String tokenValue = null;
+            String tokenSecret = null;
+            String consumerKey = null;
+            Authorization authorization = null;
+            String user = null;
+            String verifier = null;
+            Long creationTime = null;
+            Long timeToLive = null;
+            URI callback = null;
+            Map<String, String> properties = null;
+            ServiceProviderToken.Session session = null;
+
+            while (reader.hasMoreChildren()) {
+                reader.moveDown();
+                String name = reader.getNodeName();
+                String value = reader.getValue();
+                switch (name) {
+                    case ACCESS_TOKEN:
+                        accessToken = Boolean.valueOf(value);
+                        break;
+                    case TOKEN_VALUE:
+                        tokenValue = decrypt(value);
+                        break;
+                    case TOKEN_SECRET:
+                        tokenSecret = decrypt(value);
+                        break;
+                    case CONSUMER_KEY:
+                        consumerKey = value;
+                        break;
+                    case CALLBACK:
+                        callback = ofNullable(value).map(URI::create).orElse(null);
+                        break;
+                    case AUTHORIZATION:
+                        authorization = Authorization.valueOf(value);
+                        break;
+                    case USER:
+                        user = value;
+                        break;
+                    case VERIFIER:
+                        verifier = decrypt(value);
+                        break;
+                    case TOKEN_CREATION_TIME:
+                        creationTime = Long.valueOf(value);
+                        break;
+                    case TOKEN_TIME_TO_LIVE:
+                        timeToLive = Long.valueOf(value);
+                        break;
+                    case SESSION:
+                        session = unmarshalSession(reader);
+                        break;
+                    case PROPERTIES:
+                        properties = unmarshalProperties(reader);
+                }
+                reader.moveUp();
+            }
+
+            try {
+                ServiceProviderToken.ServiceProviderTokenBuilder tokenBuilder = (accessToken ?
```

Comment:
This was briefly discussed a while ago (on Gaurav's PR IIRC): I think a single `accessToken` boolean will be sufficient since there are no other cases (token is either access or is request).
I'm still not convinced the enum adds any value at all, even inside the Token class.